pkg/parquetconverter/parquet_converter_ring_test.go

```diff
@@ -0,0 +1,210 @@
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/thanos-io/objstore"
+
+	"github.com/grafana/mimir/pkg/storage/bucket"
+	"github.com/grafana/mimir/pkg/storage/tsdb/block"
+	mimir_testutil "github.com/grafana/mimir/pkg/storage/tsdb/testutil"
+	"github.com/grafana/mimir/pkg/util/test"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+// mockBlockConverter is a mock implementation of blockConverter that always succeeds.
+type mockBlockConverter struct{}
+
+// ConvertBlock implements the blockConverter interface and always returns nil (success).
+func (m mockBlockConverter) ConvertBlock(ctx context.Context, meta *block.Meta, localBlockDir string, bkt objstore.Bucket, logger log.Logger) error {
+	return nil
+}
+
+// prepareWithMockConverter creates a ParquetConverter with a mock block converter for testing.
+func prepareWithMockConverter(t *testing.T, cfg Config, bucketClient objstore.Bucket) (*ParquetConverter, *prometheus.Registry) {
+	var limits validation.Limits
+	flagext.DefaultValues(&limits)
+	overrides := validation.NewOverrides(limits, nil)
+
+	cfg.DataDir = t.TempDir()
+
+	logs := &concurrency.SyncBuffer{}
+	registry := prometheus.NewRegistry()
+
+	bucketClientFactory := func(ctx context.Context) (objstore.Bucket, error) {
+		return bucketClient, nil
+	}
+	c, err := newParquetConverter(cfg, log.NewLogfmtLogger(logs), registry, bucketClientFactory, overrides, mockBlockConverter{})
+	require.NoError(t, err)
+
+	return c, registry
+}
+
+// TestParquetConverter_InitialSyncWithWaitRing tests the parquet converter start case.
+// When several parquet-converters start up at once, we expect each one to take conversion
+// ownership of a subset of all blocks.
+func TestParquetConverter_InitialSyncWithWaitRing(t *testing.T) {
+	test.VerifyNoLeak(t)
+
+	const numUsers = 2
+	const numBlocks = 8
+
+	// We run the test with 1, 2, and 3 converters to ensure that the sharding works correctly
+	for numConverters := 1; numConverters <= 3; numConverters++ {
+		testName := fmt.Sprintf("%d converter%s", numConverters, map[bool]string{true: "s", false: ""}[numConverters > 1])
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
+			t.Cleanup(cancel)
+
+			// Create isolated bucket and blocks for this test case
+			bucketClientOnDisk, storageDir := mimir_testutil.PrepareFilesystemBucket(t)
+			now := time.Now()
+			mockTSDB(t, path.Join(storageDir, "user-1"), 24, numBlocks/2, now.Add(-24*time.Hour).Unix()*1000, now.Unix()*1000)
+			mockTSDB(t, path.Join(storageDir, "user-2"), 24, numBlocks/2, now.Add(-24*time.Hour).Unix()*1000, now.Unix()*1000)
+
+			// Uploading the blocks to in memory storage - otherwise we run out of file descriptors during the test
+			// The default limit for fsd is 1024 on linux.
```

Comment:
Did this really happen for this test? :open_mouth:  Sounds like a bug somewhere, we should investigate